### PR TITLE
[ticket/17333] UX Enhancements for Web Push Subscribing

### PR DESCRIPTION
**Pull request by iMattPro**
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket:

https://tracker.phpbb.com/browse/PHPBB-17333

Screenshot 2024-06-04 at 10 24 33 AM

### CODE REVIEW

**Comment by iMattPro** [comment]
@marc1706 This currently sets these new configs to false by default. Thus retaining the current user experience. But, should they be on by default instead? Having both these enabled does improve the discoverability of the push notification feature.

**Comment by marc1706** [comment]
I'd go with yes by default and of course only showing it once push notifications have been enabled.

**Comment by iMattPro** [comment]
> I'd go with yes by default and of course only showing it once push notifications have been enabled.

LOL. I forgot I already set these vars to be enabled by default. And ofc they don't show if web push is disabled. But I did take this opportunity to optimize the code a little so that it won't really won't execute any extra stuff if web push is disabled in general.

**Comment by marc1706** [comment]
> > I'd go with yes by default and of course only showing it once push notifications have been enabled.
> 
> LOL. I forgot I already set these vars to be enabled by default. And ofc they don't show if web push is disabled. But I did take this opportunity to optimize the code a little so that it won't really won't execute any extra stuff if web push is disabled in general.
> 
> Failing test is a scam LOL

I've restarted the test :)

File: phpBB/includes/acp/acp_board.php [hunk 1]
```diff
@@ -494,6 +494,8 @@ function main($id, $mode)
 						'webpush_enable'			=> ['lang' => 'WEBPUSH_ENABLE', 'validate' => 'bool', 'type' => 'custom', 'method' => 'webpush_enable', 'explain' => true],
 						'webpush_vapid_public'		=> ['lang' => 'WEBPUSH_VAPID_PUBLIC', 'validate' => 'string', 'type' => 'text:25:255', 'explain' => true],
 						'webpush_vapid_private'		=> ['lang' => 'WEBPUSH_VAPID_PRIVATE', 'validate' => 'string', 'type' => 'password:25:255', 'explain' => true],
+						'webpush_method_enables'	=> ['lang' => 'WEBPUSH_METHOD_ENABLES', 'validate' => 'bool', 'type' => 'radio:yes_no', 'explain' => true],
```
Comment (marc1706) [inline phpBB/includes/acp/acp_board.php:497]: How about:
```suggestion
						'webpush_method_default_enable'	=> ['lang' => 'WEBPUSH_METHOD_DEFAULT_ENABLE', 'validate' => 'bool', 'type' => 'radio:yes_no', 'explain' => true],
```

File: phpBB/includes/functions.php [hunk 2]
```diff
@@ -3869,10 +3872,24 @@ function page_header($page_title = '', $display_online_list = false, $item_id =
 		{
 			$template->assign_block_vars('notifications', $notification->prepare_for_display());
 		}
+
+		// Assign web push template vars globally (if not done already by ucp_notifications) for the dropdown subscribe button
+		if ($config['webpush_enable']
+			&& $config['webpush_dropdown_subscribe']
+			&& $template->retrieve_var('NOTIFICATIONS_WEBPUSH_ENABLE') === null)
```
Comment (marc1706) [inline phpBB/includes/functions.php:3879]:
```suggestion
		if ($config['webpush_enable'] && $config['webpush_dropdown_subscribe']
			&& $template->retrieve_var('NOTIFICATIONS_WEBPUSH_ENABLE') === null)
```

File: phpBB/includes/functions.php [hunk 3]
```diff
@@ -3869,10 +3872,24 @@ function page_header($page_title = '', $display_online_list = false, $item_id =
...
+		{
+			$methods = $phpbb_notifications->get_subscription_methods();
+			$webpush = $methods['notification.method.webpush'] ?? null;
+
+			if ($webpush)
+			{
+				$formHelper = $phpbb_container->get('form_helper');
```
Comment (marc1706) [inline phpBB/includes/functions.php:3886]:
```suggestion
				$form_helper = $phpbb_container->get('form_helper');
```

File: phpBB/includes/functions.php [hunk 4]
```diff
@@ -3869,10 +3872,24 @@ function page_header($page_title = '', $display_online_list = false, $item_id =
...
+				$template_ary = $webpush['method']->get_ucp_template_data($controller_helper, $formHelper);
```
Comment (marc1706) [inline phpBB/includes/functions.php:3887]: No need for an extra variable, just stick it directly to assign_vars ;)

**Review by marc1706 — CHANGES_REQUESTED** [review]

**Review by marc1706 — APPROVED** [review]

**Pull request merged**